### PR TITLE
Emit default ray payload access qualifiers for HLSL SM 6.7+

### LINKED ISSUE #10267: HLSL codegen emits [raypayload] without payload access qualifiers for SM 6.7+
## Summary

When compiling a ray tracing shader to DXIL with SM 6.7+, Slang emits the `[raypayload]` attribute on payload structs but does not emit the required payload access qualifiers on the struct fields. This causes DXC to reject the generated HLSL.

## Repro

**Input shader** (`test_raypayload.slang`):

```hlsl
RWTexture2D<float4> outputImage;
RaytracingAccelerationStructure topLevelAS;

struct RayPayload
{
    float3 color;
};

[shader("raygeneration")]
void rgen_main()
{
    RayDesc ray;
    ray.Origin    = float3(0, 0, 0);
    ray.Direction = float3(0, 0, 1);
    ray.TMin      = 0.001;
    ray.TMax      = 1000.0;

    RayPayload payload;
    payload.color = float3(0, 0, 0);

    TraceRay(topLevelAS, RAY_FLAG_NONE, 0xFF, 0, 0, 0, ray, payload);

    outputImage[DispatchRaysIndex().xy] = float4(payload.color, 1.0);
}
```

**Command:**

```
slangc test_raypayload.slang -entry rgen_main -stage raygeneration -target dxil -profile sm_6_8
```

**Error:**

```
dxc 1.9: test_raypayload.slang(14): error: payload type 'RayPayload_0' requires that all fields carry payload access qualifiers.
```

## Notes

- SM 6.6 and below: Slang does **not** emit `[raypayload]`, and DXC compiles successfully.
- SM 6.7+: Slang emits `[raypayload]` but omits the field-level access qualifiers.
- Tested with Slang 2026.3.1.

### PR DESCRIPTION
**Pull request by jkwak-work**
## Summary

Fixes #10267.

HLSL Shader Model 6.7+ requires every member of a `[raypayload]` struct to declare `: read(stages…) : write(stages…)` payload access qualifiers (PAQs). When a plain Slang struct (no explicit `[raypayload]` attribute) is passed to `TraceRay`, the HLSL legalization pass was already emitting the `[raypayload]` attribute on the struct but was not emitting the required PAQs on its fields, causing DXC to reject the generated HLSL.

Changes:
- **`slang-ir-hlsl-legalize.cpp`**: Add `addDefaultPayloadAccessQualifiersToField` / `addDefaultPayloadAccessQualifiersToStruct` helpers that inject default `read`/`write` qualifiers (all four stages: caller, anyhit, closesthit, miss) on fields that carry no explicit PAQ. Called from the existing `ForceVarIntoRayPayloadStructTemporarily` legalization path. Fields with any existing qualifier are left unchanged to preserve user intent.
- **`slang-emit-hlsl.cpp`**: Guard PAQ emission behind `_shouldEmitPayloadAccessQualifiers()` so access qualifiers are only emitted for DX profiles ≥ SM 6.7. Uses the cached `m_effectiveProfile` member for efficiency.
- **`tests/hlsl/raypayload-auto-paq.slang`**: Regression test verifying:
  - SM 6.8 output includes `[raypayload]` and default PAQs on all fields
  - SM 6.6 output omits the attribute and qualifiers entirely
  - Fields with explicit PAQs are preserved as-is (mixed-qualifier struct)
  - DXIL targets validate that DXC accepts the generated HLSL end-to-end

## Test plan
- [ ] `tests/hlsl/raypayload-auto-paq.slang` — HLSL and DXIL filecheck targets (SM 6.8 and SM 6.6)
- [ ] Existing `tests/hlsl/raypayload-attribute-paq.slang` — explicit `[raypayload]` struct unaffected
- [ ] Existing `tests/hlsl/raypayload-attribute.slang` — SM 6.6 struct unaffected

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/hlsl/raypayload-auto-paq.slang [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+// HLSL68: struct [raypayload] RayPayload_0
+// HLSL68: float3 color_0 : read(caller, anyhit, closesthit, miss) : write(caller, anyhit, closesthit, miss);
+// HLSL68: float3 normal_0 : read(caller, anyhit, closesthit, miss) : write(caller, anyhit, closesthit, miss);
+// HLSL66-NOT: [raypayload]
+// HLSL66: struct RayPayload_0
+// HLSL66: float3 color_0;
+// HLSL66: float3 normal_0;
+// HLSL66-NOT: : read(
+// DXIL68: define void @
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply tests/hlsl/raypayload-auto-paq.slang]: Addressed in commit 159c11133. The test now uses a separate explicit `[raypayload]` struct with non-default `read(caller) : write(miss)` qualifiers to verify preservation, keeps the implicit payload field checking default PAQs, avoids unstable struct suffixes in FileCheck, and also checks SM 6.6 does not emit `: write(`. Validation: `build/Debug/bin/slang-test.exe tests/hlsl/raypayload-auto-paq.slang` passed all 4 subtests; direct `slangc.exe` HLSL checks for SM 6.8 and SM 6.6 matched the expected PAQ behavior.

File: tests/hlsl/raypayload-auto-paq.slang [hunk 2]
```diff
@@ -0,0 +1,60 @@
+// Regression test for:
+// https://github.com/shader-slang/slang/issues/10267
+
+//TEST:SIMPLE(filecheck=HLSL68): -target hlsl -profile sm_6_8 -stage raygeneration -entry rgen_main
+//TEST:SIMPLE(filecheck=HLSL66): -target hlsl -profile sm_6_6 -stage raygeneration -entry rgen_main
+//TEST:SIMPLE(filecheck=DXIL68): -target dxil -profile sm_6_8 -stage raygeneration -entry rgen_main
+//TEST:SIMPLE(filecheck=DXIL66): -target dxil -profile sm_6_6 -stage raygeneration -entry rgen_main
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply tests/hlsl/raypayload-auto-paq.slang:6]: Addressed in commit dc82757c5. The regression test now includes explicit HLSL coverage for the SM 6.7 boundary and the canonical lib_6_7 raygeneration path, with the same assertions that implicit payloads get default PAQs and explicit payload qualifiers are preserved. Validation: build/Debug/bin/slang-test.exe tests/hlsl/raypayload-auto-paq.slang passed all 6 subtests.

File: tests/hlsl/raypayload-auto-paq.slang [hunk 3]
```diff
@@ -2,6 +2,8 @@
 // https://github.com/shader-slang/slang/issues/10267
 
 //TEST:SIMPLE(filecheck=HLSL68): -target hlsl -profile sm_6_8 -stage raygeneration -entry rgen_main
+//TEST:SIMPLE(filecheck=HLSL67): -target hlsl -profile sm_6_7 -stage raygeneration -entry rgen_main
+//TEST:SIMPLE(filecheck=HLSLLIB): -target hlsl -profile lib_6_7 -stage raygeneration -entry rgen_main
```
Comment (jkwak-work) [inline tests/hlsl/raypayload-auto-paq.slang]: This should be DXIL test:
```
//TEST:SIMPLE(filecheck=DXIL67): -target dxil -profile sm_6_7 -stage raygeneration -entry rgen_main
```

Comment (jkwak-work) [reply tests/hlsl/raypayload-auto-paq.slang]: Added in commit a7cb22d6f. The test now has a DXIL sm_6_7 boundary case with dx.dxrPayloadAnnotations assertions, while keeping the HLSL lib_6_7 coverage from the previous bot feedback. Validation: build/Debug/bin/slang-test.exe tests/hlsl/raypayload-auto-paq.slang passed all 7 subtests. Please resolve this thread if the change looks good.

File: tests/hlsl/raypayload-auto-paq.slang [hunk 4]
```diff
@@ -0,0 +1,73 @@
+// Regression test for:
+// https://github.com/shader-slang/slang/issues/10267
+
+//TEST:SIMPLE(filecheck=HLSL68): -target hlsl -profile sm_6_8 -stage raygeneration -entry rgen_main
+//TEST:SIMPLE(filecheck=HLSL67): -target hlsl -profile sm_6_7 -stage raygeneration -entry rgen_main
+//TEST:SIMPLE(filecheck=DXIL67): -target dxil -profile sm_6_7 -stage raygeneration -entry rgen_main
+//TEST:SIMPLE(filecheck=HLSLLIB): -target hlsl -profile lib_6_7 -stage raygeneration -entry rgen_main
```
Comment (jkwak-work) [inline tests/hlsl/raypayload-auto-paq.slang]: We don't need lib_6_7.
And please sort them by HLSL first and sm_6_X in ascending order.

Comment (jkwak-work) [reply tests/hlsl/raypayload-auto-paq.slang]: Addressed in commit 74137d9a9. Removed the lib_6_7 case and sorted the directives/check blocks with HLSL sm_6_6, sm_6_7, sm_6_8 first, followed by DXIL sm_6_6, sm_6_7, sm_6_8. Validation: build/Debug/bin/slang-test.exe tests/hlsl/raypayload-auto-paq.slang passed 6/6 subtests. Please resolve this thread if the change looks good.

File: tests/hlsl/raypayload-auto-paq.slang [hunk 5]
```diff
@@ -0,0 +1,68 @@
+// Regression test for:
+// https://github.com/shader-slang/slang/issues/10267
+
+//TEST:SIMPLE(filecheck=HLSL66): -target hlsl -profile sm_6_6 -stage raygeneration -entry rgen_main
+//TEST:SIMPLE(filecheck=HLSL67): -target hlsl -profile sm_6_7 -stage raygeneration -entry rgen_main
+//TEST:SIMPLE(filecheck=HLSL68): -target hlsl -profile sm_6_8 -stage raygeneration -entry rgen_main
+//TEST:SIMPLE(filecheck=DXIL66): -target dxil -profile sm_6_6 -stage raygeneration -entry rgen_main
+//TEST:SIMPLE(filecheck=DXIL67): -target dxil -profile sm_6_7 -stage raygeneration -entry rgen_main
+//TEST:SIMPLE(filecheck=DXIL68): -target dxil -profile sm_6_8 -stage raygeneration -entry rgen_main
```
Comment (jkwak-work) [inline tests/hlsl/raypayload-auto-paq.slang]: We don't need HLSL68 and DXIL68 because it is same to HLSL67 and DXIL67.
```
//TEST:SIMPLE(filecheck=HLSL67): -target hlsl -profile sm_6_8 -stage raygeneration -entry rgen_main
//TEST:SIMPLE(filecheck=DXIL67): -target dxil -profile sm_6_8 -stage raygeneration -entry rgen_main
```

Comment (jkwak-work) [reply tests/hlsl/raypayload-auto-paq.slang]: Addressed in commit 20432ebd0. Removed the redundant HLSL68 and DXIL68 test directives and their FileCheck blocks, keeping the SM 6.6 off and SM 6.7 on coverage for HLSL and DXIL. Validation: build/Debug/bin/slang-test.exe tests/hlsl/raypayload-auto-paq.slang passed 4/4 subtests. Please resolve this thread if the change looks good.

File: source/slang/slang-ir-hlsl-legalize.cpp [hunk 6]
```diff
@@ -12,6 +12,47 @@
 namespace Slang
 {
 
+static void addDefaultPayloadAccessQualifiersToField(IRBuilder& builder, IRStructKey* fieldKey)
+{
+    if (fieldKey->findDecoration<IRStageReadAccessDecoration>() ||
+        fieldKey->findDecoration<IRStageWriteAccessDecoration>())
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply source/slang/slang-ir-hlsl-legalize.cpp]: Addressed in commit d0fe8bc9f. `addDefaultPayloadAccessQualifiersToField` now checks read/write decorations independently and adds only the missing default qualifier, while preserving any existing qualifier. Source-level half-qualified implicit payloads are rejected before IR lowering, but this makes the legalization helper robust for partially decorated IR fields. Validation: `cmake.exe --build --preset debug --target slangc` and `build/Debug/bin/slang-test.exe tests/hlsl/raypayload-auto-paq.slang` passed 5/5 subtests.

File: source/slang/slang-ir-hlsl-legalize.cpp [hunk 7]
```diff
@@ -12,6 +12,47 @@
...
+    {
+        return;
+    }
+
+    IRInst* stageNames[] = {
+        builder.getStringValue(UnownedStringSlice("caller")),
+        builder.getStringValue(UnownedStringSlice("anyhit")),
+        builder.getStringValue(UnownedStringSlice("closesthit")),
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply source/slang/slang-ir-hlsl-legalize.cpp:25]: Addressed in commit d0fe8bc9f. Empty ray payload dummy fields now go through the shared default PAQ helper, so `_slang_dummy` gets the same caller/anyhit/closesthit/miss read and write defaults as non-empty implicit payload fields. The regression test now covers this emitted dummy field. Validation: `cmake.exe --build --preset debug --target slangc` and `build/Debug/bin/slang-test.exe tests/hlsl/raypayload-auto-paq.slang` passed 5/5 subtests.

File: source/slang/slang-ir-hlsl-legalize.cpp [hunk 8]
```diff
@@ -62,13 +108,16 @@ void searchChildrenForForceVarIntoStructTemporarily(IRModule* module, IRInst* in
...
                     auto structType = builder.createStructType();
-                    StringBuilder structName;
                     builder.addNameHintDecoration(structType, UnownedStringSlice(typeNameHint));
                     if (isForcedRayPayloadStruct)
-                        builder.addRayPayloadDecoration(structType);
+                        addRayPayloadDecorationIfNeeded(builder, structType);
 
                     auto elementBufferKey = builder.createStructKey();
                     builder.addNameHintDecoration(elementBufferKey, UnownedStringSlice("data"));
+                    if (isForcedRayPayloadStruct)
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply source/slang/slang-ir-hlsl-legalize.cpp:123]: Addressed in commit d0fe8bc9f. Added an `rgen_nonstruct` SM 6.7 HLSL regression case that exercises the synthesized `RayPayload_t` wrapper and checks the generated `data` field has default read/write PAQs. Validation: `cmake.exe --build --preset debug --target slangc` and `build/Debug/bin/slang-test.exe tests/hlsl/raypayload-auto-paq.slang` passed 5/5 subtests.

File: tests/hlsl/raypayload-auto-paq.slang [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+// HLSL66-DAG: float3 normal_0;
+// HLSL66-NOT: : read(
+// HLSL66-NOT: : write(
+// HLSL67-DAG: struct [raypayload] RayPayload
+// HLSL67-DAG: float3 color_0 : read(caller, anyhit, closesthit, miss) : write(caller, anyhit, closesthit, miss);
+// HLSL67-DAG: struct [raypayload] ExplicitRayPayload
+// HLSL67-DAG: float3 normal_0 : read(caller) : write(miss);
+// DXIL66: define void @
+// DXIL66-NOT: !dx.dxrPayloadAnnotations
+// DXIL67: define void @
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply tests/hlsl/raypayload-auto-paq.slang:36]: Addressed in commit d0fe8bc9f. The HLSL FileCheck coverage now matches the exact emitted field lines for the SM 6.7 implicit payload, explicit qualifier preservation, empty payload dummy field, and non-struct wrapper field, while keeping the DXIL 6.6/6.7 metadata presence gate. The `dxil` SIMPLE output is raw DXIL binary in this test harness, so the stable field-level assertions are made on the HLSL output that feeds DXC. Validation: `cmake.exe --build --preset debug --target slangc` and `build/Debug/bin/slang-test.exe tests/hlsl/raypayload-auto-paq.slang` passed 5/5 subtests.

File: source/slang/slang-emit-hlsl.cpp [hunk 10]
```diff
@@ -2133,10 +2133,13 @@ void HLSLSourceEmitter::emitSemanticsImpl(IRInst* inst, bool allowOffsets)
...
-    if (auto readAccessSemantic = inst->findDecoration<IRStageReadAccessDecoration>())
-        _emitStageAccessSemantic(readAccessSemantic, "read");
-    if (auto writeAccessSemantic = inst->findDecoration<IRStageWriteAccessDecoration>())
-        _emitStageAccessSemantic(writeAccessSemantic, "write");
+    if (_shouldEmitPayloadAccessQualifiers())
+    {
+        if (auto readAccessSemantic = inst->findDecoration<IRStageReadAccessDecoration>())
+            _emitStageAccessSemantic(readAccessSemantic, "read");
+        if (auto writeAccessSemantic = inst->findDecoration<IRStageWriteAccessDecoration>())
+            _emitStageAccessSemantic(writeAccessSemantic, "write");
```
[the remark this answers is not in the document]
Comment (jkwak-work) [reply source/slang/slang-emit-hlsl.cpp:2141]: Addressed in commit aa1b28490. Added `PartialRayPayload` to this regression test. The SM 6.6 HLSL check now covers explicit user-authored PAQs and verifies the emitted fields have no `read`/`write` syntax. The SM 6.7 check preserves the authored side and fills only the missing side, so the behavior is pinned without adding a new diagnostic. Validation: `build/Debug/bin/slang-test.exe tests/hlsl/raypayload-auto-paq.slang` passed 5/5 subtests.

**Review by szihs — APPROVED** [review]
